### PR TITLE
Scrapper, Isakov Danila - 21FPL2

### PR DESCRIPTION
**Pull request by vchemsmisl**

### CODE REVIEW

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+        self._timeout = self.config.timeout
+        self._should_verify_certificate = self.config.should_verify_certificate
+        self._headless_mode = self.config.headless_mode
 
     def _extract_config_content(self) -> ConfigDTO:
         """
         Returns config values
         """
-        pass
+        with open(self.path_to_config, 'r', encoding='utf-8') as path:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:81]: duplicated read

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+            self.config_dict = json.load(path)
+        return ConfigDTO(self.config_dict['seed_urls'],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: try dictionary unpacking for initialization of DTO

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+                                self.config_dict['should_verify_certificate'],
+                                self.config_dict['headless_mode'])
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters
         are not corrupt
         """
-        pass
+        with open(self.path_to_config, 'r', encoding='utf-8') as path:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: in init:

1. you need to extract config's content with a function
2. run validation using DTO from the previuos step
3. fill attributes of config and continue work

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+            self.config_dict = json.load(path)
+        if not isinstance(self.config_dict['seed_urls'], list):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: rework validation to operate on top of DTO instance

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+            raise IncorrectSeedURLError('seed URL is not a list')
+        regex = re.compile(r'https?://w?w?w?.')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: do you think it is possible to have for example just one w?

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+        for url in self.config_dict['seed_urls']:
+            if not isinstance(url, str):
+                raise IncorrectSeedURLError('seed URL is not str')
+            if not re.match(regex, url):
+                raise IncorrectSeedURLError('seed URL does not match standard pattern "https?://w?w?w?."')
+        if not isinstance(self.config_dict['total_articles_to_find_and_parse'], int) or \
+                self.config_dict['total_articles_to_find_and_parse'] <= 0:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: all constants should be taken from constants

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+        return self._headless_mode
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+    return requests.get(url, headers=config.get_headers())
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: you also need to set encoding to response

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
-        pass
+        return self._headless_mode
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:79]: insert random timeouts among requests

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -13,70 +35,112 @@ def __init__(self, path_to_config: Path) -> None:
...
+    return requests.get(url, headers=config.get_headers())
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:169]: provide timeout also as argument to request

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -90,25 +154,39 @@ def __init__(self, config: Config) -> None:
...
-        pass
+        art_num = 0
+        for url in self._seed_urls:
+            response = make_request(url, self.config)
+            if response.status_code != 200:
+                continue
+            main_bs = BeautifulSoup(response.text, 'lxml')
+            feed_lines = main_bs.find_all('a', {'class': 'd-block mb-0'})
+            for line in feed_lines:
+                self.urls.append(self._extract_url(line))
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: you add everything that _extract_url returns, does it guarantee to always return valid URL?

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -90,25 +154,39 @@ def __init__(self, config: Config) -> None:
...
+        self.config = config
+        self.urls = []
+        self._seed_urls = self.config.get_seed_urls()
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        return article_bs.get('href')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: your page does not contain irrelevant links?

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -90,25 +154,39 @@ def __init__(self, config: Config) -> None:
...
+        art_num = 0
+        for url in self._seed_urls:
+            response = make_request(url, self.config)
+            if response.status_code != 200:
+                continue
+            main_bs = BeautifulSoup(response.text, 'lxml')
+            feed_lines = main_bs.find_all('a', {'class': 'd-block mb-0'})
+            for line in feed_lines:
+                self.urls.append(self._extract_url(line))
+                art_num += 1
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: you always know number of collected urls as they are stored in collection. therefore, no need in additional counter

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        article_list = article.find_all('p')
+        paragraphs = [par.text for par in article_list]
+        self.article.text = '\n'.join(paragraphs)
 
     def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> None:
         """
         Finds meta information of article
         """
-        pass
+        article = article_soup.find_all('div', {'itemprop': 'headline'})[0]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: if you are sure there is only one element, use find instead of find_all

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            self.article.author = [author.text for author in authors]
+        else:
+            self.article.author = ['NOT FOUND']
+        try:
+            article_tags = article_soup.find_all('ul', {'itemprop': 'keywords'})[0]
+            article_tags_li = article_tags.find_all('li')
+            self.article.topics = [tag.find_all('a')[0].text for tag in article_tags_li]
+        except IndexError:
+            self.article.topics = []
+        article_date = article_soup.find_all('span', {'class': 'news-datetime mb-10 mr-20'})[0]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: is there a better way to find date? sequence of classes is fragile.

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         """
         Unifies date format
         """
-        pass
+        date_list = date_str.split()
+        date_list[1] = date_list[1][:3]
+        if date_list[1] == 'мая':
+            date_list[1] = 'май'
+        date_str = ' '.join(date_list)
+        locale.setlocale(category=locale.LC_ALL, locale='Russian')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: using locale is dangerous. rework to mapping as we did during lecture

File: lab_5_scrapper/scrapper.py [hunk 16]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
 
 
 def prepare_environment(base_path: Union[Path, str]) -> None:
     """
     Creates ASSETS_PATH folder if no created and removes existing folder
     """
-    pass
+    if base_path.exists():
+        shutil.rmtree(base_path)
+    new_path = Path(__file__).parent / base_path
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: base path comes as absolute, do not append it to another absolute path

File: lab_5_scrapper/scrapper.py [hunk 17]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+    new_path.mkdir(exist_ok=True, parents=True)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: exist ok is unnecessary as you just removed it

File: lab_5_scrapper/scrapper_config.json [hunk 18]
```diff
@@ -1,9 +1,9 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://www.interfax-russia.ru/volga/news"],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:2]: better layout

File: lab_5_scrapper/scrapper.py [hunk 19]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         Unifies date format
         """
-        pass
+        date_list = date_str.split()
+        date_list[1] = date_list[1][:3]
+        if date_list[1] == 'мая':
+            date_list[1] = 'май'
+        date_str = ' '.join(date_list)
+        locale.setlocale(category=locale.LC_ALL, locale='Russian')
+        return datetime.datetime.strptime(date_str, '%d %b %Y г. %H:%M')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: there is a better date in meta tag close to this one. better to parse from there

File: lab_5_scrapper/scrapper.py [hunk 20]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
 def main() -> None:
     """
     Entrypoint for scrapper module
     """
-    pass
+    prepare_environment(constants.ASSETS_PATH)
+    configuration = Config(path_to_config=constants.CRAWLER_CONFIG_PATH)
+    crawler = Crawler(config=configuration)
+    crawler.find_articles()
+    search_urls = crawler.urls
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: unnecessary variable

File: lab_5_scrapper/scrapper_config.json [hunk 21]
```diff
@@ -1,9 +1,9 @@
...
+    "headers": {"user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/111.0"},
+    "total_articles_to_find_and_parse": 100,
+    "encoding": "utf8",
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:15]: do you use it anywhere?

File: requirements.txt [hunk 22]
```diff
@@ -0,0 +1,2 @@
+requests
```
Comment (demid5111) [inline requirements.txt]: specify particular versions (!)

File: requirements.txt [hunk 23]
```diff
@@ -0,0 +1,2 @@
...
+bs4
```
Comment (demid5111) [inline requirements.txt]: check out CI status

File: lab_5_scrapper/scrapper.py [hunk 24]
```diff
@@ -120,45 +198,89 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        article_authors = article_soup.find_all('span', {'itemprop': 'author'})[0]
+        authors = article_authors.find_all('span', {'itemprop': 'name'})
+        if authors:
+            self.article.author = [author.text for author in authors]
+        else:
+            self.article.author = ['NOT FOUND']
+        try:
+            article_tags = article_soup.find_all('ul', {'itemprop': 'keywords'})[0]
+            article_tags_li = article_tags.find_all('li')
+            self.article.topics = [tag.find_all('a')[0].text for tag in article_tags_li]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: get text from li tags is not enough?

**Review by demid5111 — CHANGES_REQUESTED** [review]
nice progress, keep going

File: lab_5_scrapper/scrapper.py (hunk 20 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py:169]: Done

File: lab_5_scrapper/scrapper.py (hunk 12 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 15 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: requirements.txt (hunk 22 above)
Comment (vchemsmisl) [reply requirements.txt]: Done

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py:81]: Done

File: lab_5_scrapper/scrapper.py (hunk 8 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py:79]: Done

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 11 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 17 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 16 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper_config.json (hunk 21 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper_config.json:15]: Now I do, so Done

File: lab_5_scrapper/scrapper.py (hunk 14 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done - I used itemprop

File: lab_5_scrapper/scrapper.py (hunk 13 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 19 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 24 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper_config.json (hunk 18 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper_config.json:2]: Done

File: lab_5_scrapper/scrapper.py [hunk 25]
```diff
@@ -1,8 +1,62 @@
...
+from core_utils.article.article import Article
+from core_utils.article.io import to_meta, to_raw
+from core_utils.config_dto import ConfigDTO
+from core_utils.constants import (ASSETS_PATH, CRAWLER_CONFIG_PATH,
+                                  NUM_ARTICLES_UPPER_LIMIT,
+                                  TIMEOUT_LOWER_LIMIT, TIMEOUT_UPPER_LIMIT)
+
+
+class IncorrectSeedURLError(TypeError):
+    '''
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: use triple double quotes in docstrings

File: lab_5_scrapper/scrapper.py [hunk 26]
```diff
@@ -13,70 +67,110 @@ def __init__(self, path_to_config: Path) -> None:
         """
         Initializes an instance of the Config class
         """
-        pass
+        self.path_to_config = path_to_config
+        self.config = Config._extract_config_content(self)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: why not self._extract_config_content()?

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: you can add here empty string, right? is it correct?

File: lab_5_scrapper/scrapper.py [hunk 27]
```diff
@@ -90,25 +184,63 @@ def __init__(self, config: Config) -> None:
...
+        for url in self._seed_urls:
+            response = make_request(f'{url}?per-page={num_arts}', self.config)
+            if response.status_code != 200:
+                continue
+            main_bs = BeautifulSoup(response.text, 'lxml')
+            feed_lines = main_bs.find_all('a', {'class': 'd-block mb-0'})
+            for line in feed_lines[:num_arts]:
+                self.urls.append(self._extract_url(line))
+
+        # WAS JUST TRYING ANOTHER WAY OF DYNAMIC SITE CRAWLING, I'LL DELETE IT LATER
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: remove unused code

Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: dynamic scrapping is not required to get mark 10, do not invest much effort unless it is your hobby

File: lab_5_scrapper/scrapper.py [hunk 28]
```diff
@@ -120,45 +252,88 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         """
-        pass
+        return datetime.datetime.strptime(date_str, '%Y-%m-%dT%H:%M:%S%z')
 
     def parse(self) -> Union[Article, bool, list]:
         """
         Parses each article
         """
-        pass
+        response = requests.get(self.full_url,
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: use make_request function

File: requirements.txt [hunk 29]
```diff
@@ -0,0 +1,3 @@
+bs4==0.0.1
```
Comment (demid5111) [inline requirements.txt]: name of the library should be different

**Review by demid5111 — CHANGES_REQUESTED** [review]
nice, continue

File: lab_5_scrapper/scrapper.py (hunk 27 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 25 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 26 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 28 above)
Comment (vchemsmisl) [reply lab_5_scrapper/scrapper.py]: Done

File: requirements.txt (hunk 29 above)
Comment (vchemsmisl) [reply requirements.txt]: Done

**Comment by demid5111** [comment]
remove main2() from main

**Comment by demid5111** [comment]
nothing to review to this moment.

**Comment by vitazaytseva** [comment]
Well done! You can proceed to work on Lab 6)

**Pull request closed without merging**